feat/new-lessons-on-js-promises (#36045)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/create-a-javascript-promise.english.md

---
id: 5cdafbb0291309899753167f
title: Create a Javascript Promise
challengeType: 1
---

## Description
<section id='description'>
A promise in javascript is exactly what it sounds like. You use it to make a promise to do something, usually asynchronous. When the task completes you either fulfill your promise or fail to do so. It’s a constructor function, so they are created with the <code>new</code> keyword. It needs a function as its argument with two parameters, <code>resolve</code> and <code>reject</code>. These are methods used to determine the outcome of the promise. The syntax looks like this:

```js
const myPromise = new Promise((resolve, reject) => {

});
```

</section>

## Instructions
<section id='instructions'>
Create a new promise called <code>makeServerRequest</code>. Pass in a function with <code>resolve</code> and <code>reject</code> parameters to the promise.
</section>

## Tests
<section id='tests'>

```yml
tests:
  - text: Your promise should be in a variable called <code>makeServerRequest</code>.
    testString: assert(typeof(makeServerRequest) === "object" && code.match(/(let|var|const)\s*makeServerRequest\s*=\s*new\s*Promise((\s|\w)*)/g));

>   - text: Your promise should be passed a function with <code>resolve</code> and <code>reject</code> as parameters.

Reviewer: ```suggestion
  - text: Your promise should receive a function with <code>resolve</code> and <code>reject</code> as parameters.
```

Reviewer: I think my suggestion is still better. @RandellDawson what do you say?

---
id: 5cdafbb0291309899753167f
title: Create a Javascript Promise
challengeType: 1
---

## Description
<section id='description'>

> A promise in javascript is exactly what it sounds like. You use it to make a promise to do something, usually asynchronous. When the task completes you either fulfill your promise or fail to do so. It’s a constructor function, so they are created with the <code>new</code> keyword. It needs a function as its argument with two parameters, <code>resolve</code> and <code>reject</code>. These are methods used to determine the outcome of the promise. The syntax looks like this:

Reviewer 2: That's a good point.  I like your new version more, but I'm getting "You do not have permission to push to this repository." again...

> A promise in javascript is exactly what it sounds like. You use it to make a promise to do something, usually asynchronous. When the task completes you either fulfill your promise or fail to do so. It’s a constructor function, so they are created with the <code>new</code> keyword. It needs a function as its argument with two parameters, <code>resolve</code> and <code>reject</code>. These are methods used to determine the outcome of the promise. The syntax looks like this:

Author: I always get that message

> A promise in javascript is exactly what it sounds like. You use it to make a promise to do something, usually asynchronous. When the task completes you either fulfill your promise or fail to do so. It’s a constructor function, so they are created with the <code>new</code> keyword. It needs a function as its argument with two parameters, <code>resolve</code> and <code>reject</code>. These are methods used to determine the outcome of the promise. The syntax looks like this:

Author: I pretty much always get that message

> A promise in JavaScript is exactly what it sounds like. You use it to make a promise to do something, usually asynchronously. When the task completes you either fulfill your promise or fail to do so. <code>Promise</code> is a constructor function, so you need to use the <code>new</code> keyword to create one. It takes a function as its argument with two parameters, <code>resolve</code> and <code>reject</code>. These are methods used to determine the outcome of the promise. The syntax looks like this:

Reviewer 3: ```suggestion
A promise in JavaScript is exactly what it sounds like - you use it to make a promise to do something, usually asynchronously. When the task completes, you either fulfill your promise or fail to do so. <code>Promise</code> is a constructor function, so you need to use the <code>new</code> keyword to create one. It takes a function, as its argument, with two parameters - <code>resolve</code> and <code>reject</code>. These are methods used to determine the outcome of the promise. The syntax looks like this:
```

Author: Why would we want to use hyphens here? That can't be correct grammar/punctuation. I added one of the commas. I didn't think the others were needed. Let me know if you still think they should be added.

Reviewer 3: I added a hyphen here so as to answer what the first sentence means. 

You said "A promise in JavaScript is exactly what it sounds like". So, **what does it sound like**? Without a further explanation of what it sounds like, the sentence is incomplete by itself. The hyphen is intended to complete the sentence, by adding the second sentence to complete "what it sounds like".

```js
const myPromise = new Promise((resolve, reject) => {

});
```

</section>

## Instructions
<section id='instructions'>
Create a new promise called <code>makeServerRequest</code>. Pass in a function with <code>resolve</code> and <code>reject</code> parameters to the constructor.
</section>

## Tests
<section id='tests'>

```yml
tests:

>   - text: Your promise should be in a variable called <code>makeServerRequest</code>.

Reviewer 4: I notice if you leave off the variable declaration (i.e. `var`, `const`, `let`) the first test fails, but since the "variable not defined" error in the browser console does not show up (as it really should), it might be a little confusing to what the actually problem is.  I think the following change in the test will make it clearer.
```suggestion
  - text: You should assign a promise to a declared variable named <code>makeServerRequest</code>.
```
